AppActionVisualizer/app/src/main/java/com/example/appactionvisualizer/utils/Utils.java

```diff
@@ -19,4 +49,40 @@ private static Toast getToast(String message, Context context) {
   private static Toast getToast(String message, Context context, int length) {
     return Toast.makeText(context, message, length);
   }
+
+  //go to play store
+  public static void jumpToStore(Context context, final String packageName) {
+    context.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=" + packageName)));
+  }
+
+  //first check if the url can jump to activity, if not, check if package exists
+  public static void jumpToApp(Context context, final String url, final String packageName) {
+    try {
+      Intent intent = Intent.parseUri(url, 0);
+      context.startActivity(intent);
+    } catch (Exception e) {
+      try {
```

Comment:
Instead of having logic in catch() block, could you do a check before firing the intent?
if (intent.resolveActivity(getPackageManager()) != null) {
        startActivity(intent);
    }

Reply:
I tried this way; However, it seems like in some cases the intent cannot find the corresponding activity though we've installed the app. Checking the activity could not indicate whether we've installed the app or not.
I rewrote the logic as below, would this be better?
```
  //first check if package exists since user may haven't installed the app
  //then check if the url can jump to activity
    try {
      context.getPackageManager().getApplicationInfo(packageName, 0);
    } catch (PackageManager.NameNotFoundException e) {
      Utils.jumpToStore(context, packageName);
    }
    try {
      Intent intent = Intent.parseUri(url, 0);
      context.startActivity(intent);
    } catch (Exception e) {
      String errorMsg;
      if(e instanceof URISyntaxException) {
        errorMsg = context.getString(R.string.error_parsing);
      }else if(e instanceof ActivityNotFoundException) {
        errorMsg = context.getString(R.string.error_activity);
      }else{
        errorMsg = context.getString(R.string.error_unknown);
      }
      Utils.showMsg(errorMsg, context);
    }
```